<a href="https://colab.research.google.com/github/thurmboris/Data-Science_4_Sustainability/blob/main/01_Python-Basics/01_Python-Basics_1_Variables-Operators-Conditionals.ipynb" target="_blank" rel="noopener"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python Basics: let's start!

<img src='https://www.agent-x.com.au/wp-content/uploads/2011/06/Perfect-Programmer-dfe194b-e8d3b11-b960bd5.jpg' width="400">

Source: [Agent-X Comics - Perfect Programming](https://www.agent-x.com.au/comic/perfect-programming/)

## Contents

In this notebook, we will start programming with Python: we will discover how to store data in "variables" and manipulate them using "operators". Finally, we will assess whether individuals are cooperating or not using "conditionals". Let our journey begin!      

- [Variables](#Variables)
- [Data Type](#Data_Type)
  - [Text data](#Text_data)
  - [Numeric data](#Numeric_data)
  - [Casting and type conversion](#casting)
- [Operators](#Operators)
  - [Arithmetic Operators](#arithmetic-operators)
    - [Operations on integers](#operating-int)
    - [Operations on floats](#operating-float)
    - [Operations on strings](#operating-str)
    - [Order of operations](#operating-order)
  - [Assignment Operators](#assignment-operators)
  - [Comparison Operators and Boolean](#comparison-operators)
  - [Identity Operators](#identity)
  - [Logical Operators](#logical)
- [Conditionals](#Conditionals)

## Variables <a name="Variables"></a>

Whether you are programming in Python or pretty much any other language, you will be working with **variables**. Variables are containers for storing data values. We will talk more about **objects** later, but a variable, like everything in Python, is an object. The following can be properties of a variable:
1. The **type** of variable. E.g., is it an integer, like `2`, or a string, like `'Hello, world.'`?
2. The **value** of the variable.
Depending on the type of the variable, you can do different things to it and other variables of similar type.

A variable is created the moment you first assign a value to it:

In [1]:
a = 3

Notice that when you assign a value to a variable, there is not visible output. However, now if we ask for `a`, its value will be displayed:

In [2]:
a

3

Be careful, when you assign an already assigned variable, the value will be overwritten!

In [3]:
a = 7
a

7

<span style='color:blue'> **Tips:** </span>
    
- Use explicit name for your variables: you and someone reading your code should be able to understand what is the variable.
- You can - actually, should - comment your code using `#` to explain what you are doing. 

Variables names are **case-sensitive**:

In [4]:
# Defining "A" will not overwrite "a"
A = "What a wonderful day!"
print(A)
print(a)

What a wonderful day!
7


## Data Type <a name="Data_Type"></a>

Variables can store data of different types, and different types can do different things. The Python's built-in `type()` function allows to determine the type of some data/variables.

### Text data <a name="Text_data"></a>

The first type of data we have encountered is **Text**, such as `"Hello, world"`. In Python language, Text type data are called **string**. When asked about the type of `"Hello, world"`, the `type()` function will return `str`, the short version of string:

In [5]:
type("Hello world")

str

Note that there are several ways to define strings. You can use single or double quotes. For instance, `'This is a string'` and `"This is a string"` are equivalent. You can also use triple quotes to extend a string over multiple lines:

In [6]:
my_str = '''Triple quotes allows...
to extend strings over multiple lines.'''

print(my_str)

Triple quotes allows...
to extend strings over multiple lines.


### Numeric data <a name="Numeric_data"></a>

There are three numeric types in Python:
- integer `int`
- real `float`
- complex `complex`

**Integer** (**int**) is a whole number, positive or negative, without decimals, of unlimited length:

In [7]:
type(4)

int

**Float** stands for "floating point number" and is a number, positive or negative, containing one or more decimals.

In [8]:
type(3.9)

float

Note that you can also use scientific notation using `e`:

In [9]:
type(9.5e-8)

float

Finally, you can define **complex** number. Note that in Python, the imaginary part is defined by `j`:

In [10]:
type(1+2j)

complex

Be careful when you define and operate on data: `3.9` is a float, but `'3.9'` is a string!

### Casting and type conversion <a name="casting"></a>

There may be times when you want to specify a type on to a variable. This can be done with "casting", using constructor functions:
- `int()` - constructs an integer number from an integer literal, a float literal (by removing all decimals), or a string literal (providing the string represents a whole number)
- `float()` - constructs a float number from an integer literal, a float literal or a string literal (providing the string represents a float or an integer)
- `complex()` - constructs a complex number from a wide variety of data types, including strings, integer literals and float literals
- `str()` - constructs a string from a wide variety of data types, including strings, integer literals and float literals

In [11]:
cast_int = int(1.2)
cast_float = float(4)
cast_complex = complex(3.5)
cast_str = str(9.2)

print(type(cast_int), cast_int)
print(type(cast_float), cast_float)
print(type(cast_complex), cast_complex)
print(type(cast_str), cast_str)

<class 'int'> 1
<class 'float'> 4.0
<class 'complex'> (3.5+0j)
<class 'str'> 9.2


Note that when converting a `float` to an `int`, the interpreter does not round the result, but gives the floor.

In [12]:
int(2.6)

2

The `int()`, `float()`, `complex()`, and `str()` functions are very useful to convert one variable type into another. For example, often times we will import data from a text file, i.e., many strings, but we will want to perform operations on numbers:

In [13]:
imp_str = '5.3'
conv_str = float('5.3')
print(type(imp_str), imp_str)
print(type(conv_str), conv_str)

<class 'str'> 5.3
<class 'float'> 5.3


## Operators <a name="Operators"></a>

### Arithmetic operators <a name="arithmetic-operators"></a>

**Operators** allow you to do things with variables, like add them. They are represented by special symbols, like `+` and `*`. For now, we will focus on **arithmetic** operators. Python's arithmetic operators are:

|action|operator|
|:-------|:----------:|
|addition | `+`|
|subtraction | `-`|
|multiplication | `*`|
|division | `/`|
|raise to power | `**`|
|modulo | `%`|
|floor division | `//`|

**Warning**: Do not use the `^` operator to raise to a power. That is actually the operator for bitwise XOR, which we will not cover for now.

#### Operations on integers <a name="operating-int"></a>

Let's try the arithmetic operators on integers:

In [14]:
3+5

8

In [15]:
3-5

-2

In [16]:
3*5

15

In [17]:
3/5

0.6

In [18]:
3**5

243

In [19]:
3%5

3

In [20]:
3//5

0

Notice that `3/5` produces a `float`, even though `3` and `5` are `int`: 

In [21]:
print(type(3+5))
print(type(3/5))

<class 'int'>
<class 'float'>


Note than you cannot divide by zero. If you do so, you will get an error message that explicitly tells you what went wrong:

In [22]:
7/0

ZeroDivisionError: division by zero

#### Operations on floats <a name="operating-float"></a>

Let's now try the arithmetic operators on floats:

In [23]:
2.1 + 3.2

5.300000000000001

Wait a minute!  We know `2.1 + 3.2 = 5.3`, but Python gives `5.300000000000001`. This is due to the fact that floating point numbers are stored with a finite number of binary bits. There will always be some rounding errors. This means that as far as the computer is concerned, it cannot tell you that `2.1 + 3.2` and `5.3` are equal. This is important to remember when dealing with floats, as we will see later.

In [24]:
# Very very close to zero because of finite precision
5.3 - (2.1 + 3.2)

-8.881784197001252e-16

In [25]:
2.1-3-2

-2.9

In [26]:
2.1*3.2

6.720000000000001

In [27]:
2.1/3.2

0.65625

In [28]:
2.1**3.2

10.74241047739471

In [29]:
2.1%3.2

2.1

In [30]:
2.1//3.2

0.0

Everything works as expected aside from the floating point precision previously mentioned. As before, you cannot divide by zero:

In [31]:
7.4/0.0

ZeroDivisionError: float division by zero

Note that you can operate on integers and floats. In other words, you do not need to convert integers into floats to perform mixed operations:

In [32]:
1+6.8

7.8

#### Operations on strings <a name="operating-str"></a>

Finally let's try some of these operations on strings. Yes, we will perform mathematical operations on strings! What will we get? Well, let's see...

In [33]:
'We can '+'add strings!'

'We can add strings!'

The result is intuitive: adding strings together concatenates them! How about subtracting strings?

In [34]:
'Can we '-'subtract strings?'

TypeError: unsupported operand type(s) for -: 'str' and 'str'

Ah, too bad, we cannot subtract strings. Well, it actually makes sense, subtracting strings would be weird. At least, we got a nice error message explaining us that the `str` and `str` are unsupported operant types for the `-` operation. 

Similarly, we cannot perform multiplication, raising of power, etc., with two strings. How about multiplying a string by an integer?

In [35]:
'cat '*3

'cat cat cat '

Wow, three `'cat '`! It makes sense: multiplication by an integer is the same thing as just adding multiple times, so the Python interpreter concatenates the string several times.

#### Order of operations <a name="operating-order"></a>

The order of operations follows common convention. Exponentiation comes first, followed by multiplication and division, floor division, and modulo. Next comes addition and subtraction. In order of precedence, our arithmetic operator table is

|precedence|operators|
|:-------:|:----------:|
|1 | `**`|
|2 | `*`, `/`, `//`, `%`|
|3 | `+`, `-`|

You can also group operations with parentheses. Operations within parentheses are always evaluated first.

<span style='color:blue'> **Tips:** </span> *do not* use excessive parentheses. Excessive parentheses makes your code less readable, and can lead to mistakes. Trust the order of operations ;)

In [36]:
1**3 + 2**3 + 3**3 + 4**3 + 5**3

225

In [37]:
(1+2+3+4+5)**2

225

Wooow! The sum of the cubes of 1, 2, ..., 5 is equal to the square of the sum from 1 to 5. Can you demonstrate that this property is true for all *n*?

### Assignment operators <a name="assignment-operators"></a>

Assignment operators are used to assign values to variables. We have already encountered one of them: that's right, the  `=` operator that allows to initiate a variable.

In [38]:
var = 7
print(type(var), var)

<class 'int'> 7


Now, let's say we want to update the value of our variable `var`. As previously mentioned, you can directly overwrite a variable. You can also use operations. For instance, suppose we want to add `3.9` to `var`. You can use the `+` operator:

In [39]:
var = var+3.9
print(type(var), var)

<class 'float'> 10.9


Notice that we changed the type of our variable `var` from an `int` to a `float`.

Instead of using the `+` arithmetic operator to update our variable, there was a more efficient way, using the assignment operator `+=`:

In [40]:
var = 7
var+= 3.9
print(var)

10.9


The `+=` operator told the interpreter to take the value of `var` and add `3.9` to it, changing the type of `var` in the intuitive way if need be. 

Similarly, the other arithmetic operators have similar assignment operators:

|Operator|Example|Same as|
|:-------:|:----------:|:----------:|
|`=` | `var = 7` | `var = 7` |  
|`+=` | `var += 7` | `var = var + 7`| 
|`-=` | `var -= 7` | `var = var - 7`|
|`*=` | `var *= 7` | `var = var * 7`|
|`/=` | `var /= 7` | `var = var / 7`|
|`**=` | `var **= 7` | `var = var ** 7`|
|`%=` | `var %= 7` | `var = var % 7`| 
|`//=` | `var //= 7` | `var = var // 7`|

### Comparison operators and Boolean <a name="comparison-operators"></a>

**Comparison operators** (also called **relational operators**) are used to compare two values.

Let's start by assessing if two values are equal. We use the `==` operator:

In [41]:
8 == 8

True

In [42]:
8==9

False

Wow! Python confirmed that 8 is equal to 8 but is not equal to 9!

Wait a minute, we know what "True" and "False" mean in English, i.e., words that indicate truth. We can guess they have the same meaning in Python. But what is their type? After all, we have so far seen `str`, `int`, `float`, and `complex` data types. Are `True` and `False` strings? No! `True` and `False` have a special type, called `bool`, short for **Boolean**.

In [43]:
print(type(True))
print(type(False))

<class 'bool'>
<class 'bool'>


The name "Boolean" comes from the English mathematician and philosopher [Georges Boole](https://en.wikipedia.org/wiki/George_Boole)

<img src='https://upload.wikimedia.org/wikipedia/commons/c/ce/George_Boole_color.jpg' width="200">

In Boolean logic, a statement can be either true or false, and the Boolean are associated with numerical value: `True`has the value `1`, and `False` has the value `0`:

In [44]:
True == 1

True

In [45]:
False == 0

True

You can even perform arithmetic operations on boolean. The result will be an `int`:

In [46]:
sum_bool = True + False

print(type(sum_bool), sum_bool)

<class 'int'> 1


Ok, now that we understand what boolean are, let's test it with some floats:

In [47]:
5.3 == 5.3 

True

As expected. One more time:

In [48]:
2.1+3.2 == 5.3

False

As expect... Wait, what?! How come `2.1 + 3.2` is not `5.3`? Well, remember, there was rounding errors when summing `2.1` and `3.2`. This is the floating point arithmetic issue. Note that floating point numbers that can be exactly represented with binary numbers do not have this problem:

In [49]:
2.2+3.2 == 5.4

True

Unfortunately, this behavior is unpredictable, so **never use the `==` operator with `float`**.

Comparison is not restricted to equality. Here are the other comparison operators:

|English|Python|
|:-------|:----------:|
|is equal to | `==`|
|is not equal to | `!=`|
|is greater than | `>`|
|is less than | `<`|
|is greater than or equal to | `>=`|
|is less than or equal to | `<=`|

Let's try them!

In [50]:
-1 > 6

False

In [51]:
4 <= 4

True

We can even chain comparison operators:

In [52]:
1<2<3

True

However, even if it is legal, do not mix the direction of the comparison operators: 

In [53]:
1 < 3 > 2

True

See, chaining comparison operators check the relation element-by-element. In the above example, it means that `1` and `2` are not compared. 

Finally, we can use comparison operators on strings:

In [54]:
'Federer' > 'Nadal'

False

Wait, what?! Python got crazy! I mean, Python has never seen a tennis match so how does it compares tennis players anyway? Well, it does not. It actually compares the characters of strings. 

How so? In Python, characters are encoded with [Unicode](https://en.wikipedia.org/wiki/Unicode). This is a standardized library of characters from many languages around the world that contains over 100,000 characters. Each character has a unique number associated with it. We can access what number is assigned to a character using Python's built-in `ord()` function.

In [55]:
ord('a')

97

The relational operators on characters compare the values that the `ord` function returns. So, using a relational operator on `'a'` and `'b'` means you are comparing `ord('a')` and `ord('b')`. When comparing strings, the interpreter first compares the first character of each string. If they are equal, it compares the second character, and so on. So, the reason that `'Federer' > 'Nadal'` gives a value of `False` is because `ord('F') < ord('N')`. We're safe, but the debate is still not settled...

Note that a result of this scheme is that testing for equality of strings means that **all** characters must be equal. This is the most common use case for relational operators with strings.

### Identity operators <a name="identity"></a>

**Identity operators** are used to compare objects, not if they are equal, but if they are actually the same object, with the same memory location. The two identity operators are:

|English|Python|
|:-------|:----------:|
|is the same object | **`is`**|
|is not the same object | **`is not`**|

That's right. The operators are pretty much the same as English! Let's see these operators in action and get at the difference between `==` and `is`. Let's use the **`is`** operator to investigate how Python stored variables in memory, starting with `float`s.

In [56]:
a = 6.1
b = 6.1

a == b, a is b

(True, False)

See, `a` and `b` have the same value so the `==` operators returns `True`. However, they are not the same object because they are stored in different places in memory, so the `is` operator returns `False`. 

They can occupy the same place in memory if we do a `b = a` assignment:

In [57]:
a = 6.1
b = a

a == b, a is b

(True, True)

Because we assigned `b = a`, they necessarily have the same (immutable) value. The two variables also occupy the same place in memory for efficiency. Thus, both `==` and `is` operators return `True`.

However, if we reassign the value of `a`, then the interpreter is placing `a` in a new space in memory, so `a` and `b` are not longer the same object:

In [58]:
a = 6.1
b = a
a = 8.5

a == b, a is b

(False, False)

The same discussion is valid for most `int` and `str`. Why most and not all?

For integers between between `-5` and `256`, Python employs **integer caching**, meaning that these integers will occupy the same space in memory. This caching does not happen for more negative or larger integers:

In [59]:
a = 93
b = 93
c = 708
d = 708

a is b, c is d

(True, False)

Similarly, Python is sometimes doing [**string interning**](https://en.wikipedia.org/wiki/String_interning) which allows for (sometimes very) efficient string processing. Whether two strings occupy the same place in memory depends on what the strings are:

In [60]:
a = 'Hello'
b = 'Hello'
c = 'Hello world!'
d = 'Hello world!'

a is b, c is d

(True, False)

You generally do not need to worry about caching and interning for **immutable** variables. Immutable means that once the variables are created, their values cannot be changed. If we do change the value the variable gets a new place in memory. All variables we've encountered so far (`int`, `float`, `complex` and `str`) are immutable.

### Logical operators <a name="logical"></a>

**Logical operators** can be used to connect relational and identity operators. Python has three logical operators.

|Logic|Python|
|:-------|:----------:|
|AND | `and`|
|OR | `or`|
|NOT | `not`|

The `and` operator means that if both operands are `True`, return `True`. For instance, if statement "P" is true and statement "Q" is true, then the statement "P `and` Q" is also true. However, if either "P" or "Q" is false, then "P `and` Q" is false.

On the other hand, the `or` operator gives `True` if *either* of the operands are `True`.

|P|Q|P `and` Q|P `or` Q|
|:--:|:--:|:--:|:--:|
|`True` | `True` |`True` |`True` |
|`True` | `False`|`False`|`True` |
|`False`|`True` |`False`|`True` |
|`False`|`False`|`False`|`False`|

Let's try!

In [61]:
True and True

True

In [62]:
True and False

False

In [63]:
False and False

False

In [64]:
True or False

True

Finally, the `not` operator negates the logical result:

In [65]:
not False

True

In [66]:
not False and True

True

In [67]:
not (True and False)

True

Note that it is important to specify the ordering of your operations, particularly when using the `not` operator.

Note also that

    a < b < c
    
is equivalent to

    (a < b) and (b < c)

With these new types of operators in hand, we can construct a more complete table of operator precedence.

|precedence|operators|
|:-------|:----------:|
|1 | `**`|
|2 | `*`, `/`, `//`, `%`|
|3 | `+`, `-`|
|4 | `<`, `>`, `<=`, `>=`|
|5 | `==`, `!=`|
|6 | `=`, `+=`, `-=`, `*=`, `/=`, `**=`, `%=`, `//=`|
|7 | `is`, `is not`|
|8 | `and`, `or`, `not`|

## Conditionals <a name="Conditionals"></a>

**Conditionals** are used to tell your computer to do a set of instructions depending on whether or not a Boolean is `True`. In other words, we are telling the computer:

    if something is true:
        do task a
    otherwise:
        do task b

In fact, the syntax in Python is almost exactly the same. As always, an example speaks volumes. 

We are going to study the condition for cooperation in [collective action problem](https://en.wikipedia.org/wiki/Collective_action_problem), also called social dilemma. In such situation, all individuals would be better off cooperating but fail to do so because of conflicting interests between them, which discourage joint action. The illustration below represents a social dilemma:
- the blue line tells us what individuals get - i.e., their payoffs - when cooperating.  
- the red-dashed line tells us what individuals get when they do not cooperate, which is called "defecting".
- the payoffs depend on how many people cooperate in the population, represented by the x-axis. When the share of cooperators increase, the payoffs of individuals also increaset
- it is costly to cooperate: the payoffs when defecting (red-dashed line) is larger than the payoffs when cooperating (blue line). The difference between the two lines is called the "individual cost" to cooperate.
- finally, we call the social benefit the difference in payoffs between a situation of full cooperation and a situation of full defection. 

<img src='https://i.postimg.cc/44HkDp79/Social-Dilemma.png' width="800">

Many environmental issues takes the form of a social dilemma. For example, recycling requires time and efforts but decreases the consumption of materials if widely adopted, which decrease the extraction of materials and the associated pollution. Purchasing an electric vehicle is costly but reduces air pollutants, associated with various respiratory and cardio-vascular health diseases, and greenhouse gas emissions, responsible for climate change.

We will assume that individuals have *homo moralis* preferences: they consider not only their selfish payoff but also what happens when all others do the same action. The weight of selfishness and morality depends on the individual degree of morality. Recent economic literature has demonstrated that such preference provides an evolutionary advantage (see e.g., Alger & Weibull, 2013). In other words, a population of *homo moralis* outperforms any other type of preferences, such as the fully-selfish *homo oeconomicus*.

We can demonstrate that *homo moralis* individuals cooperate in a social dilemma (e.g., perform a pro-environmental action) when their social benefit weighted by their degree of morality is greater than their individual cost of acting weighted by their degree of selfishness (Ayoubi and Thurm, 2021). 

*References:*
- Alger, I., & Weibull, J. W. (2013). Homo moralis—preference evolution under incomplete information and assortative matching. Econometrica, 81(6), 2269-2302. [DOI: 10.3982/ECTA10637](https://doi.org/10.3982/ECTA10637)
- Ayoubi, C. and Thurm, B. (2021). Pro-environmental behavior and morality: An economic model with heterogeneous preferences. Available at [SSRN 3681953](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=3681953)

Ok, enough words, let's assess whether a given *homo moralis* cooperates.

In [68]:
cost = 1            # individual cost 
benefit = 3         # social benefit
kappa = 0.5         # degree of morality

# condition for cooperation: 
# social benefit times kappa is greater than individual cost times (1-kappa)

if benefit*kappa >= cost*(1-kappa):
    print('The individual cooperates!')

The individual cooperates!


Youhouuu, good news for nature, the individual performs an environmental-friendly action!

Now, let's review the syntax of the `if` statement. The Boolean expression, `benefit*kappa >= cost*(1-kappa)`, is called the **condition**. If it is `True`, the indented statement below is executed. In this case, we print the string `'The individual cooperates!'`. Also, do not forget the `:` at the end of the `if` statement!

This brings up a very important aspect of Python syntax: <span style="color: dodgerblue; font-weight: bold;">
Indentation matters.
</span> Any lines with the same level of indentation will be evaluated together.

In [69]:
if benefit*kappa >= cost*(1-kappa):
    print('The individual cooperates!')
    print('Same level of intentation, so still printed!')

The individual cooperates!
Same level of intentation, so still printed!


What happens if the condition is `False`? Let's try with an individual with degree of morality `kappa=0`, i.e., the infamous fully-selfish *homo oeconomicus*:

In [70]:
kappa = 0                # degree of morality

# condition for cooperation: 
# social benefit times kappa is greater than individual cost times (1-kappa)

if benefit*kappa >= cost*(1-kappa):
    print('The individual cooperates!')

Nothing happened. This is because we did not tell Python what to do if the condition was evaluated as `False`. We can add that with an `else` **clause** in the conditional.

In [71]:
kappa = 0      # degree of morality

# condition for cooperation: 
# social benefit times kappa is greater than individual cost times (1-kappa)

if benefit*kappa >= cost*(1-kappa):
    print('The individual cooperates!')
else:
    print('What a shame, the individual does not cooperate...')

What a shame, the individual does not cooperate...


We can assess several conditions by using an `elif` clause. For example, say we have two individuals, Edoardo and Quentin, and we want to check if they both cooperates, if only one of them cooperates, or if they both do not care for the environment. We will combine the logical operator `and` with the comparison operators `>=` and `<`:

In [72]:
kappa_1 = 0.2     # degree of morality of the first individual
kappa_2 = 0.3     # degree of morality of the second individual

# condition for cooperation: 
# social benefit times kappa is greater than individual cost times (1-kappa)

if benefit*kappa_1 >= cost*(1-kappa_1) and benefit*kappa_2 >= cost*(1-kappa_2):
    print('Both individual cooperates!')
elif benefit*kappa_1 < cost*(1-kappa_1) and benefit*kappa_2 < cost*(1-kappa_2):
    print('What a shame, nobody cooperates...')
else:
    print('Only one individual cooperates')

Only one individual cooperates


It seems like either Edoardo or Quentin are not cooperating... Besides the degree of morality, there are many other factors influencing the cooperation between individuals. In the notebook "Practice-with-Python", you can explore how the level of knowledge of environmental issues influence cooperation. Time to practice! 